ssl/s3_enc.c

```diff
@@ -273,20 +273,46 @@ int ssl3_finish_mac(SSL_CONNECTION *s, const unsigned char *buf, size_t len, int
          * we calculate the digest when initiating s->s3.handshake_dgst at which
          * point we know what the protocol version is.
          */
+        if (s->negotiated_version == DTLS1_3_VERSION) {
+           /*
+            * In DTLS 1.3 we need to parse the messages that are buffered to
+            * be able to remove message_sequence, fragment_size and fragment_offset
+            * from the Transcript Hash calculation.
+            */
+            while (len > 0) {
+                PACKET hmhdr;
+                unsigned long hmbodylen;
+                unsigned int msgtype;
+                size_t hmhdrlen;
+
+                if (!ossl_assert(len >= SSL3_HM_HEADER_LENGTH)
+                    || !PACKET_buf_init(&hmhdr, buf, SSL3_HM_HEADER_LENGTH)
+                    || !PACKET_get_1(&hmhdr, &msgtype)
+                    || !PACKET_get_net_3(&hmhdr, &hmbodylen)) {
+                    SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+                    return 0;
+                }
 
-        if (SSL_CONNECTION_IS_DTLS13(s) && hmhdr_incl
-                && ossl_assert(len >= DTLS1_HM_HEADER_LENGTH)) {
-            ret = EVP_DigestUpdate(s->s3.handshake_dgst, buf, SSL3_HM_HEADER_LENGTH);
-            ret &= EVP_DigestUpdate(s->s3.handshake_dgst,
-                                    buf + DTLS1_HM_HEADER_LENGTH,
-                                    len - DTLS1_HM_HEADER_LENGTH);
-        } else {
-            ret = EVP_DigestUpdate(s->s3.handshake_dgst, buf, len);
-        }
+                if (msgtype == SSL3_MT_MESSAGE_HASH)
+                    hmhdrlen = SSL3_HM_HEADER_LENGTH;
+                else
+                    hmhdrlen = DTLS1_HM_HEADER_LENGTH;
 
-        if (!ret) {
-            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
-            return 0;
+                if (!ossl_assert(hmhdrlen + hmbodylen <= len)
+                    || !EVP_DigestUpdate(s->s3.handshake_dgst, buf, SSL3_HM_HEADER_LENGTH)
+                    || !EVP_DigestUpdate(s->s3.handshake_dgst, buf + hmhdrlen, hmbodylen)) {
+                    SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+                    return 0;
+                }
```

Comment:
Perhaps a comment here, to the effect that the first `SSL3_HM_HEADER_LENGT` bytes of the header are the content that we want to include in the checksum and the remaining `DTLS1_HM_HEADER_LENGTH - SSL3_HM_HEADER_LENGTH` bytes (if any) are the extra fragmentation fields that are excluded from the checksum)